packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart

```diff
@@ -0,0 +1,111 @@
+
+import '../../src/common.dart';
+import '../../src/context.dart';
+
+void main() {
+  group('Android migration', () {
+    
+    testWithoutContext('migrators succeed', () {
+      final FakeAndroidMigrator fakeAndroidMigrator = FakeAndroidMigrator();
+      final ProjectMigration migration = ProjectMigration(<ProjectMigrator>[fakeAndroidMigrator]);
+      migration.run();
+    });
+
+    group('migrate the Gradle "clean" task to lazy declaration', () {
+      late MemoryFileSystem memoryFileSystem;
+      late BufferLogger bufferLogger;
+      late FakeAndroidProject project;
+      late File topLevelGradleBuildFile;
+
+      setUp(() {
+        memoryFileSystem = MemoryFileSystem();
+        bufferLogger = BufferLogger.test();
+        project = FakeAndroidProject();
+        project.hostAppGradleRoot = memoryFileSystem.currentDirectory;
+        topLevelGradleBuildFile = project.hostAppGradleRoot.childFile('build.gradle');
+      });
+
+      testUsingContext('skipped if files are missing', () {
+        final TopLevelGradleBuildFileMigration androidProjectMigration = TopLevelGradleBuildFileMigration(
+          project,
+          bufferLogger,
+        );
+        androidProjectMigration.migrate();
+        expect(topLevelGradleBuildFile.existsSync(), isFalse);
+
+        expect(bufferLogger.traceText, contains('Top-level Gradle build file not found, skipping "clean" task migration.'));
+        expect(testLogger.statusText, isEmpty);
+      });
+
+      testUsingContext('skipped if nothing to upgrade', () {
+        topLevelGradleBuildFile.writeAsStringSync('''
+tasks.register("clean", Delete) {
+  delete rootProject.buildDir
+}
+        ''');
+
+        final TopLevelGradleBuildFileMigration androidProjectMigration = TopLevelGradleBuildFileMigration(
+          project,
+          bufferLogger,
+        );
+        final DateTime topLevelGradleBuildFileLastModified = topLevelGradleBuildFile.lastModifiedSync();
+        androidProjectMigration.migrate();
+
+        expect(topLevelGradleBuildFile.lastModifiedSync(), topLevelGradleBuildFileLastModified);
+        expect(testLogger.statusText, isEmpty);
+      });
+
+      testUsingContext('top-level build.gradle is migrated', () {
+        topLevelGradleBuildFile.writeAsStringSync('''
+task clean(type: Delete) {
+    delete rootProject.buildDir
+}
+''');
+
+        final TopLevelGradleBuildFileMigration androidProjectMigration = TopLevelGradleBuildFileMigration(
+          project,
+          bufferLogger,
+        );
+        androidProjectMigration.migrate();
+
+        expect(bufferLogger.traceText, contains('Migrating "clean" Gradle task to lazy declaration style.'));
+        expect(testLogger.statusText, isEmpty);
+
+        expect(topLevelGradleBuildFile.readAsStringSync(), equals('''
+tasks.register("clean", Delete) {
+    delete rootProject.buildDir
+}
+'''));
+      });
+    });
+  });
+}
+
+class FakeAndroidProject extends Fake implements AndroidProject {
```

Comment:
Another `FakeAndroidProject` class already exists [here](https://github.com/flutter/flutter/blob/cd19817d3fdbd89646710a4f09fcd9b2acfe8a0b/packages/flutter_tools/test/general.shard/plugins_test.dart#L1805).

I think it's okay to duplicate it because it looks like they exist for different purposes, but I'd rather ask you :)

Reply:
I agree. It's not even really duplicating so much as it is two different tests using the same functionality, and making up the same reasonable name for their local private class that uses it.

(It's not private through the Dart underscore-prefix mechanism, but it's nevertheless effectively private because it's defined in a file that defines `main` and won't be imported.)